File: clients/python/llm_prefill_decode_speed_visualizer_api_client/api/default/parse_constraints.py
from http import HTTPStatus
from typing import Any, cast
from urllib.parse import quote

import httpx

from ...client import AuthenticatedClient, Client
from ...types import Response, UNSET
from ... import errors

from ...models.parse_constraints_response_200 import ParseConstraintsResponse200
from ...models.problem import Problem
from typing import cast



def _get_kwargs(
    *,
    q: str,

) -> dict[str, Any]:
    

    

    params: dict[str, Any] = {}

    params["q"] = q


    params = {k: v for k, v in params.items() if v is not UNSET and v is not None}


    _kwargs: dict[str, Any] = {
        "method": "get",
        "url": "/api/parse-constraints",
        "params": params,
    }


    return _kwargs



def _parse_response(*, client: AuthenticatedClient | Client, response: httpx.Response) -> ParseConstraintsResponse200 | Problem | None:
    if response.status_code == 200:
        response_200 = ParseConstraintsResponse200.from_dict(response.json())



        return response_200

    if response.status_code == 400:
        response_400 = Problem.from_dict(response.json())



        return response_400

    if response.status_code == 429:
        response_429 = Problem.from_dict(response.json())



        return response_429

    if client.raise_on_unexpected_status:
        raise errors.UnexpectedStatus(response.status_code, response.content)
    else:
        return None


def _build_response(*, client: AuthenticatedClient | Client, response: httpx.Response) -> Response[ParseConstraintsResponse200 | Problem]:
    return Response(
        status_code=HTTPStatus(response.status_code),
        content=response.content,
        headers=response.headers,
        parsed=_parse_response(client=client, response=response),
    )


def sync_detailed(
    *,
    client: AuthenticatedClient | Client,
    q: str,

) -> Response[ParseConstraintsResponse200 | Problem]:
    r""" Parse plain-language constraints into the canonical constraint JSON

     Converts a natural-language constraint string (e.g. \"self-hosted Qwen 27B at Q4 for 10 users under
    $1500\") into the canonical constraint struct used by /api/sizing and /api/best. Deterministic
    regex/heuristics — no external LLM calls. Returns the echoed input, the parsed struct (null = not
    stated) and an `ambiguities` array listing every assumption (e.g. \"10 users: assume 1 stream each
    or batched?\"), plus a ready-made `sizingQuery` for the downstream decision endpoint.

    Args:
        q (str):

    Raises:
        errors.UnexpectedStatus: If the server returns an undocumented status code and Client.raise_on_unexpected_status is True.
        httpx.TimeoutException: If the request takes longer than Client.timeout.

    Returns:
        Response[ParseConstraintsResponse200 | Problem]
     """


    kwargs = _get_kwargs(
        q=q,

    )

    response = client.get_httpx_client().request(
        **kwargs,
    )

    return _build_response(client=client, response=response)

def sync(
    *,
    client: AuthenticatedClient | Client,
    q: str,

) -> ParseConstraintsResponse200 | Problem | None:
    r""" Parse plain-language constraints into the canonical constraint JSON

     Converts a natural-language constraint string (e.g. \"self-hosted Qwen 27B at Q4 for 10 users under
    $1500\") into the canonical constraint struct used by /api/sizing and /api/best. Deterministic
    regex/heuristics — no external LLM calls. Returns the echoed input, the parsed struct (null = not
    stated) and an `ambiguities` array listing every assumption (e.g. \"10 users: assume 1 stream each
    or batched?\"), plus a ready-made `sizingQuery` for the downstream decision endpoint.

    Args:
        q (str):

    Raises:
        errors.UnexpectedStatus: If the server returns an undocumented status code and Client.raise_on_unexpected_status is True.
        httpx.TimeoutException: If the request takes longer than Client.timeout.

    Returns:
        ParseConstraintsResponse200 | Problem
     """


    return sync_detailed(
        client=client,
q=q,

    ).parsed

async def asyncio_detailed(
    *,
    client: AuthenticatedClient | Client,
    q: str,

) -> Response[ParseConstraintsResponse200 | Problem]:
    r""" Parse plain-language constraints into the canonical constraint JSON

     Converts a natural-language constraint string (e.g. \"self-hosted Qwen 27B at Q4 for 10 users under
    $1500\") into the canonical constraint struct used by /api/sizing and /api/best. Deterministic
    regex/heuristics — no external LLM calls. Returns the echoed input, the parsed struct (null = not
    stated) and an `ambiguities` array listing every assumption (e.g. \"10 users: assume 1 stream each
    or batched?\"), plus a ready-made `sizingQuery` for the downstream decision endpoint.

    Args:
        q (str):

    Raises:
        errors.UnexpectedStatus: If the server returns an undocumented status code and Client.raise_on_unexpected_status is True.
        httpx.TimeoutException: If the request takes longer than Client.timeout.

    Returns:
        Response[ParseConstraintsResponse200 | Problem]
     """


    kwargs = _get_kwargs(
        q=q,

    )

    response = await client.get_async_httpx_client().request(
        **kwargs
    )

    return _build_response(client=client, response=response)

async def asyncio(
    *,
    client: AuthenticatedClient | Client,
    q: str,

) -> ParseConstraintsResponse200 | Problem | None:
    r""" Parse plain-language constraints into the canonical constraint JSON

     Converts a natural-language constraint string (e.g. \"self-hosted Qwen 27B at Q4 for 10 users under
    $1500\") into the canonical constraint struct used by /api/sizing and /api/best. Deterministic
    regex/heuristics — no external LLM calls. Returns the echoed input, the parsed struct (null = not
    stated) and an `ambiguities` array listing every assumption (e.g. \"10 users: assume 1 stream each
    or batched?\"), plus a ready-made `sizingQuery` for the downstream decision endpoint.

    Args:
        q (str):

    Raises:
        errors.UnexpectedStatus: If the server returns an undocumented status code and Client.raise_on_unexpected_status is True.
        httpx.TimeoutException: If the request takes longer than Client.timeout.

    Returns:
        ParseConstraintsResponse200 | Problem
     """


    return (await asyncio_detailed(
        client=client,
q=q,

    )).parsed

File: api/_handlers/parse-constraints.js
import { parseConstraints, constraintsToSizingQuery, isCoarseQuantLabel, constraintsToBestQuery } from '../_parse_constraints.js';
import { enforceRateLimit } from '../_ratelimit.js';
import { sendJson } from '../_schema.js';
import { sendProblem } from '../_errors.js';

export const config = { runtime: 'nodejs' };

/**
 * GET /api/parse-constraints?q=<plain-language constraints>   (#65)
 * POST /api/parse-constraints  {"q": "<...>"}                 (also accepts `text`)
 *
 * Converts a natural-language constraint string —
 *   "self-hosted Qwen 27B at Q4 for 10 users under $1500"
 * — into the canonical constraint JSON used by /api/sizing and /api/best.
 * Pure regex/heuristics: deterministic, no external LLM calls. The response
 * echoes the input, returns the parsed struct (null = not stated) and an
 * `ambiguities` array for every assumption the parser had to make, so agents
 * can ask the user instead of silently guessing. `sizingQuery` carries a
 * ready-made query string for the downstream /api/sizing decision endpoint.
 */
export default function handler(req, res) {
  if (!enforceRateLimit(req, res)) return;

  if (req.method === 'OPTIONS') {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    return res.status(204).end();
  }

  const params = req.method === 'POST' ? (req.body || {}) : (req.query || {});
  const raw = params.q ?? params.text ?? params.constraints ?? '';

  if (!String(raw).trim()) {
    return sendProblem(res, req, {
      code: 'INVALID_PARAMS',
      detail: 'Missing ?q=<plain-language constraints>. Example: ?q=self-hosted%20Qwen%2027B%20at%20Q4%20for%2010%20users%20under%20%241500'
    });
  }

  const { input, constraints, ambiguities } = parseConstraints(raw);
  // #563: a bare-level quant ("Q4") matches no stored benchmark tag, so it
  // used to 404 the ready-made sizing query on its own canonical example.
  // constraintsToSizingQuery now omits coarse quants; say so explicitly so
  // agents know why ?quant= is missing from the emitted query.
  if (isCoarseQuantLabel(constraints.quantization)) {
    ambiguities.push({
      field: 'quantization',
      message: `"${constraints.quantization}" is a bare level and was omitted from sizingQuery — stored quants are full tags like q4_k_m / iq4_xs, and exact-match filtering on "${constraints.quantization}" returns zero runs. Re-add ?quant= once you know the exact variant.`
    });
  }
  const sizingParams = constraintsToSizingQuery(constraints);
  const sizingQueryString = sizingParams.toString();
  const bestParams = constraintsToBestQuery(constraints);

  return sendJson(res, {
    description: 'Natural-language constraint parsing: canonical constraint struct + explicit ambiguities. Feed sizingQuery to /api/sizing or bestQuery to /api/best for ranked hardware; resolve each ambiguity before trusting the numbers.',
    input,
    recognizedCount: Object.values(constraints).filter(v => v != null).length,
    constraints,
    ambiguities,
    ...(sizingQueryString
      ? {
          sizingQuery: `/api/sizing?${sizingQueryString}`,
          // Query-string-only form (#563): sizingQuery keeps its legacy
          // path-prefixed shape for existing consumers; this field appends
          // directly to a base URL without double-prefixing.
          sizingQueryString
        }
      : { sizingQuery: null, sizingQueryString: null }),
    ...(bestParams.toString() ? { bestQuery: `/api/best?${bestParams.toString()}` } : { bestQuery: null })  }, { cacheTtl: 3600 });
}
